src/engine/routing_algorithms/map_matching.cpp

```diff
@@ -248,8 +263,48 @@ SubMatchingList mapMatching(SearchEngineData<Algorithm> &engine_working_data,
                         current_lengths[s_prime] = network_distance;
                         current_pruned[s_prime] = false;
                         model.breakage[t] = false;
+                        current_segment_id[s_prime] = last_id;
                     }
                 }
+                // Here we can add previous states to the current viterbi if we suppose, that user
+                // can stay
+                double precision = DEFAULT_GPS_PRECISION;
+                if (!trace_gps_precision.empty())
+                    precision = (*trace_gps_precision[t]);
+                const auto distance = util::coordinate_calculation::haversineDistance(
+                    candidates_list[prev_unbroken_timestamp][s].phantom_node.location,
+                    trace_coordinates[t]);
+
+                // We can use a previous state as one of possible states if it is inside the
+                // precision range, except for the last state in trace
+                if (distance < 3 * precision && t < candidates_list.size() - 1)
+                {
+                    const PhantomNodeWithDistance &tmpNode =
+                        candidates_list[prev_unbroken_timestamp][s];
+                    if (std::any_of(candidates_list[t].begin(),
+                                    candidates_list[t].end(),
+                                    [&](const PhantomNodeWithDistance &node) {
+                                        return node.phantom_node.IsIndistinct(tmpNode.phantom_node);
+                                    }))
+                        continue;
+
+                    candidates_list[t].push_back(tmpNode);
+                    candidates_list[t].back().phantom_node.input_location = trace_coordinates[t];
+                    candidates_list[t].back().distance = distance;
+                    map_matching::EmissionLogProbability emission_log_probability(precision);
+
+                    // We don't add the transition probability because staying is prefered than
```

Comment:
if we aren't adding high probability, isn't it going to be a low-probability state, thus not preferred one?